### PR TITLE
20x fast convex_hull_image for 2D

### PR DESCRIPTION
**Pull request by AnihilatorGun**

### CODE REVIEW

File: imops/morphology.py [hunk 1]
```diff
@@ -517,3 +522,82 @@ def distance_transform_edt(
         return result[0]
 
     return None
+
+
+def convex_hull_image_2d(image, offset_coordinates=True):
```
Comment (maxme1) [inline imops/morphology.py:527]: just call it `convex_hull_image` and fail at runtime if the image is not 2D

Comment (AnihilatorGun) [reply imops/morphology.py]: Done

File: imops/morphology.py [hunk 2]
```diff
@@ -7,26 +7,26 @@
 from scipy.ndimage import distance_transform_edt as scipy_distance_transform_edt, generate_binary_structure
 from scipy.ndimage._nd_image import euclidean_feature_transform
 from scipy.spatial import ConvexHull, QhullError
+from skimage._shared.utils import warn as skimage_warn
```
Comment (maxme1) [inline imops/morphology.py:9]: why do you need skimage_warn? isn't the regular `warnings` module enough?

Comment (AnihilatorGun) [reply imops/morphology.py:9]: Just copied code from skimage, idk why they use custom warn)

Comment (maxme1) [reply imops/morphology.py:9]: let's get rid of imports from protected submodules where we can. this is one clearly can be replaced by `warnings`

Comment (AnihilatorGun) [reply imops/morphology.py]: Fair, fixed that

**Comment by vovaf709** [comment]
Fixes #64 #45

File: tests/test_convex_hull.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    image = invert(data.horse())
+
+    try:
+        chull_ref = convex_hull_image(image, offset_coordinates=offset_coordinates, include_borders=True)
+    except TypeError:
+        chull_ref = convex_hull_image(image, offset_coordinates=offset_coordinates)
+
+    chull = convex_hull_image_fast(image, offset_coordinates=offset_coordinates)
+
+    assert not (chull < image).any()
```
Comment (vovaf709) [inline tests/test_convex_hull.py:60]: `(chull >= image).all()`

File: tests/test_convex_hull.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    assert not (chull < chull_ref).any()
```
Comment (vovaf709) [inline tests/test_convex_hull.py:61]: Lets quantify potential difference somehow. Mb something like 
`assert ((chull - chull_ref).sum() / chull_ref.sum()) < threshold

File: tests/test_convex_hull.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    coords_ref = unique_rows((coords[:, None, :] + offsets).reshape(-1, 2))
+
+    coords = _offset_unique(coords)
+
+    # _left_right_bounds has another order
+    assert len(coords_ref) == len(unique_rows(np.concatenate((coords, coords_ref), 0)))
+
+
+def test_convex_hull_image(offset_coordinates):
+    image = invert(data.horse())
```
Comment (vovaf709) [inline tests/test_convex_hull.py:51]: Also lets add some random cases

File: imops/morphology.py [hunk 6]
```diff
@@ -517,3 +527,77 @@ def distance_transform_edt(
...
+    Examples
+    --------
+    ```python
+    chull = convex_hull_image(x)
+    ```
+    """
+
+    ndim = image.ndim
+    if ndim != 2:
+        raise RuntimeError(f'convex_hull_image is currently implemented only for 2D arrays, got {ndim}D array')
```
Comment (vovaf709) [inline imops/morphology.py:559]: Seems like `ValueError`

**Review by vovaf709 — APPROVED** [review]

**Review by maxme1 — APPROVED** [review]

**Pull request merged**